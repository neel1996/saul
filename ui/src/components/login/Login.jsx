import React from "react";

export default function Login() {
    return (
        <div className="flex mx-auto my-auto p-4 rounded-md bg-slate-600 shadow-lg w-1/3"></div>
    );
}
