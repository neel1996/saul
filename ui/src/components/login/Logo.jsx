import { faBolt } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import React from "react";

export default function Logo() {
    return (
        <div data-testid="logo">
            <div className="mx-auto flex align-middle items-center justify-center gap-4 select-none">
                <div className="p-4 rounded-lg shadow-md bg-gray-600 text-indigo-400">
                    <FontAwesomeIcon icon={faBolt} size="2x" />
                </div>
                <div className="block items-center">
                    <div className="font-bold text-gray-300 text-6xl">Saul</div>
                    <div className="font-sans text-sm font-semibold text-indigo-300">
                        Know what is in your document
                    </div>
                </div>
            </div>
            <div className="border-b border-gray-600 w-11/12 mx-auto my-6"></div>
        </div>
    );
}
