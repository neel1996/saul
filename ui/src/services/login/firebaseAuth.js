import { validateUser } from "@services/user/validateUser";
import {
    getAuth,
    signInWithPopup,
    GoogleAuthProvider,
    GithubAuthProvider,
} from "firebase/auth";

export const loginWithGoogle = async () => {
    const auth = getAuth();

    return await signInWithPopup(auth, new GoogleAuthProvider()).then(
        authenticatedUser
    );
};

export const loginWithGithub = async () => {
    const auth = getAuth();

    return await signInWithPopup(auth, new GithubAuthProvider()).then(
        authenticatedUser
    );
};

const authenticatedUser = async (response) => {
    const { user } = response;
    const authToken = user.accessToken;

    sessionStorage.setItem("authToken", authToken);

    const userDetails = {
        userId: user?.uid,
        name: user?.displayName,
        email: user?.email,
        avatar: user?.photoURL,
    };

    await validateUser(userDetails);

    return userDetails;
};
