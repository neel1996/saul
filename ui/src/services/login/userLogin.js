import { post } from "@root/AxiosInstance";

export const login = async (request) => {
    return await post("/login", request)
        .then((data) => {
            if (data.authToken) {
                sessionStorage.setItem("authToken", data.authToken);
                return data;
            }

            throw new Error("Invalid credentials");
        })
        .catch((error) => {
            throw new Error(error);
        });
};
