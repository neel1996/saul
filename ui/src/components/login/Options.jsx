import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faGithub, faGoogle } from "@fortawesome/free-brands-svg-icons";
import { loginWithGoogle, loginWithGithub } from "@services/login";

export default function Options() {
    const loginOptions = [
        {
            name: "google",
            icon: faGoogle,
            login: loginWithGoogle,
        },
        {
            name: "github",
            icon: faGithub,
            login: loginWithGithub,
        },
    ];

    return (
        <div data-testid="login-options" className="my-4">
            <div className="text-center text-base font-semibold text-gray-500">
                Sign in with
            </div>
            <div
                data-testid="login-options-cta"
                className="flex items-center align-middle justify-center gap-8 my-4"
            >
                {loginOptions.map((option) => (
                    <div
                        key={option.name}
                        data-testid={`login-with-${option.name}`}
                        className="my-auto p-3 text-base border border-gray-600 transition-all rounded-full shadow-lg bg-gray-600 cursor-pointer hover:shadow-sm text-indigo-400"
                        onClick={option.login}
                    >
                        <FontAwesomeIcon icon={option.icon} size="2x" />
                    </div>
                ))}
            </div>
        </div>
    );
}
