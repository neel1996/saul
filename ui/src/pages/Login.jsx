import Login from "@components/login";
import React from "react";

export default function LoginPage() {
    return <Login />;
}
