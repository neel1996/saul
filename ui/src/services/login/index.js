import { loginWithGoogle } from "./firebaseAuth";
import { loginWithGithub } from "./firebaseAuth";

export { loginWithGoogle, loginWithGithub };
