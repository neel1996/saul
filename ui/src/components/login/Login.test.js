import { loginWithGoogle, loginWithGithub } from "@services/login";
import "@testing-library/jest-dom";
import { act, fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import { redirect } from "react-router-dom";

import Login from "./Login";

jest.mock("@services/login");
jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    redirect: jest.fn(),
}));

describe("Login", () => {
    afterEach(() => {
        jest.clearAllMocks();
        jest.resetAllMocks();
    });

    it("should render login component correctly", async () => {
        const { container } = await render(<Login />);

        expect(await screen.findByTestId("login")).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    it("should login with google on click", async () => {
        loginWithGoogle.mockResolvedValue({});

        await render(<Login />);

        const loginWithGoogleButton = await screen.findByTestId(
            "login-with-google"
        );

        // eslint-disable-next-line testing-library/no-unnecessary-act
        await act(async () => {
            await fireEvent.click(loginWithGoogleButton);
        });

        expect(loginWithGoogle).toHaveBeenCalledTimes(1);
        expect(redirect).toHaveBeenCalledTimes(1);
        expect(redirect).toHaveBeenCalledWith("/");
    });

    it("should login with github on click", async () => {
        loginWithGithub.mockResolvedValue({});

        await render(<Login />);

        const loginWithGithubButton = await screen.findByTestId(
            "login-with-github"
        );

        // eslint-disable-next-line testing-library/no-unnecessary-act
        await act(async () => {
            await fireEvent.click(loginWithGithubButton);
        });

        expect(loginWithGithub).toHaveBeenCalledTimes(1);
        expect(redirect).toHaveBeenCalledTimes(1);
        expect(redirect).toHaveBeenCalledWith("/");
    });
});
