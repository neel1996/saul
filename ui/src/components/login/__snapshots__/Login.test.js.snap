// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render login component correctly 1`] = `
<div>
  <div
    class="flex mx-auto my-auto w-full h-full bg-indigo-500"
    data-testid="login"
  >
    <div
      class="flex mx-auto my-auto w-full h-full"
      style="background-image: url(); background-size: 70%; background-repeat: no-repeat; background-position: center;"
    >
      <div
        class="bg-gray-700 p-6 rounded-lg shadow-lg w-11/12 md:w-1/2 xl:w-1/3 sm:w-11/12 mx-auto my-auto border border-gray-600"
      >
        <div
          data-testid="logo"
        >
          <div
            class="mx-auto flex align-middle items-center justify-center gap-4 select-none"
          >
            <div
              class="p-4 rounded-lg shadow-md bg-gray-600 text-indigo-400"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-bolt fa-2x "
                data-icon="bolt"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M349.4 44.6c5.9-13.7 1.5-29.7-10.6-38.5s-28.6-8-39.9 1.8l-256 224c-10 8.8-13.6 22.9-8.9 35.3S50.7 288 64 288H175.5L98.6 467.4c-5.9 13.7-1.5 29.7 10.6 38.5s28.6 8 39.9-1.8l256-224c10-8.8 13.6-22.9 8.9-35.3s-16.6-20.7-30-20.7H272.5L349.4 44.6z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <div
              class="block items-center"
            >
              <div
                class="font-bold text-gray-300 text-6xl"
              >
                Saul
              </div>
              <div
                class="font-sans text-sm font-semibold text-indigo-300"
              >
                Know what is in your document
              </div>
            </div>
          </div>
          <div
            class="border-b border-gray-600 w-11/12 mx-auto my-6"
          />
        </div>
        <div
          class="my-4"
          data-testid="login-options"
        >
          <div
            class="text-center text-base font-semibold text-gray-500 select-none"
          >
            Sign in with
          </div>
          <div
            class="flex items-center align-middle justify-center gap-8 my-4"
            data-testid="login-options-cta"
          >
            <div
              class="my-auto p-3 text-base border border-gray-600 transition-all rounded-full shadow-lg bg-gray-600 cursor-pointer hover:shadow-sm text-indigo-400"
              data-testid="login-with-google"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-google fa-2x "
                data-icon="google"
                data-prefix="fab"
                focusable="false"
                role="img"
                viewBox="0 0 488 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M488 261.8C488 403.3 391.1 504 248 504 110.8 504 0 393.2 0 256S110.8 8 248 8c66.8 0 123 24.5 166.3 64.9l-67.5 64.9C258.5 52.6 94.3 116.6 94.3 256c0 86.5 69.1 156.6 153.7 156.6 98.2 0 135-70.4 140.8-106.9H248v-85.3h236.1c2.3 12.7 3.9 24.9 3.9 41.4z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <div
              class="my-auto p-3 text-base border border-gray-600 transition-all rounded-full shadow-lg bg-gray-600 cursor-pointer hover:shadow-sm text-indigo-400"
              data-testid="login-with-github"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-github fa-2x "
                data-icon="github"
                data-prefix="fab"
                focusable="false"
                role="img"
                viewBox="0 0 496 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
